### DISCUSSION TITLE
Imgur alternative?

### DISCUSSION THREAD

**Discussion by mdxe** [flameshot-org/flameshot#1929] [Q&A]
Are there any plans to offer alternate image hosting providers?

**Comment by borgmanJeremy** [comment]
In short: yes

Here is the long answer and my current design thoughts. Most info is from this RFC: https://github.com/flameshot-org/flameshot/pull/955

Edit on 2021/12/02 to refine requirements 

# Requirements
- [ ] Compatibility with ShareX [plug-in files](https://getsharex.com/docs/custom-uploader)
- [ ] Failed or timeout on upload does not crash Flameshot or cause unresponsive GUI
- [ ] Upload parser should be decoupled from main flameshot application so it can be easily tested. The actual parser should have no Qt dependencies. 
- [ ] Invalid plug-in file does not crash Flameshot, instead an error should be displayed
- [ ] Option to completely disable uploads
- [ ] Option to compile out the ability to upload a file
- [ ] Flameshot proper should have 0 knowledge of the backend. IE there should not be a case switch statement that implements 'special' code for any one platform. It either uploads using the plugin file, or does not. 
  - If a particular backend does not work with the ShareX plugin format, we should work with them as an upstream
- [ ] A successful upload shall be logged in the Upload history. Any relevant server response should be stored here as well. 

# UI Requirements
- [ ] No Change to main flameshot UI. The existing upload button will kick-off an upload to the preconfigured upload service.
- [ ] The Upload service shall be selectable from the side panel
- [ ] A new Config tab shall be created for upload services. This shall support configuration of
  - Upload Timeout
  - Path to Plugin Folder
  - Default upload service

# Testing requirements
- [ ] Clone the ShareX plugin repo and parse each file. No errors should be thrown.
- [ ] Set up a web server in a dockerfile to act as a surrogate upload platform. Confirm that each element of the plugin is properly seen on the web server side. 

# Out of Scope
- Unlike ShareX we will not support creating and editing the plugin files

**Comment by tomtastic** [comment]
[ignore]
I'd really like to see a config option to disable uploads to imgur, it's a pain point when using flameshot in a corporate environment.
[/ignore]

**Being able to configure alternate upload destinations would be a distinct bonus.**

**Reply by mdxe** [reply]
can't your firewall control internet connection on a per-app basis?  That would be much more secure.... (blacklist everything by default)

**Comment by mmahmoudian** [comment]
@tomtastic
I think you should consider these two issues:

- #1760
- #1529

**Comment by jpzone282** [comment]
Hello  is an Imgur alternative still on the Table, maybe other drive uploads such as google drive, box, Dropbox etc?

**Comment by 404invalid-user** [comment]
any updates on this? only being able to use imgur is a big letdown.

**Reply by mmahmoudian** [reply]
As said before, this would not happen within the application itself. The plan is to enable such feature as plugins via the plugin system. For now you have to use things like rclone or your own scripts, which is not a hard thing to do and many of the members of our community have already put together the scripts and have shared them, for example for Google Drive and Dropbox.

**Reply by 404invalid-user** [reply]
whats the point on this application then if i have to make my own hacky scripts? Guess ill just continue using the in built screenshot program with my script then.

**Reply by mmahmoudian** [reply]
> whats the point on this application then if i have to make my own hacky scripts? 

1. The point of this application is to take screenshot and allow the user to annotate it.

2. Then don't use hacky scripts, write proper scripts. That is on you and out of the scope of Flameshot.

> Guess ill just continue using the in built screenshot program with my script then.

By all means. Use the software that fits your workflow best. But as the maintainer of Flameshot, I can confidently say that Flameshot is not an imgur interface, and users should really understand that. In my personal opinion, even imgur integration was a mistake and out of the scope of this software.

**Reply by tomtastic** [reply]
Is there an opportunity here to remove the Imgur integration, but add it as an optional script, along with perhaps some of the better community scripts ?
It's tight integration is what stopped my org from being able to use FlameShot more widely.

**Reply by borgmanJeremy** [reply]
I would vote to remove imgur entirely.

**Reply by tessus** [reply]
> Then don't use hacky scripts, write proper scripts. That is on you and out of the scope of Flameshot.

How do I integrate these scripts with flameshot though? There is no send-to-script button (like the `Upload the selection` button) to invoke a script from flameshot. Also, just uploading it alone does not help, one has to know the resulting URL. flameshot would have to present the user with a URL (and maybe a copy to clipboard button), getting the data from stdout from the script.

I just came across this project and it is awesome. It's not a problem to save to files and then upload them, but I think having the option to do it in one go might be a huge usability plus.
However, imgur as the only integrated option is a problem when it comes to privacy and image rights. I rather use my own upload server. ;-)

I haven't read every single line of documentation yet, so please forgive my questions:

- Is a plug-in system already available or is this planned for he future? e.g. something like sharex plugin files. 
- And to reitereate my previous question: How do I integrate scripts with flameshot?

**Reply by mmahmoudian** [reply]
@tessus I'm also a fan of your work in Joplin :) Thanks for your contributions in that FLOSS project.

> How do I integrate these scripts with flameshot though? There is no send-to-script button (like the Upload the selection button) to invoke a script from flameshot. Also, just uploading it alone does not help, one has to know the resulting URL. flameshot would have to present the user with a URL (and maybe a copy to clipboard button), getting the data from stdout from the script.

Flameshot has [a very good CLI](https://flameshot.org/docs/advanced/commandline-options/) for macOS and Linux (Windows CLI is not there yet). So one can for example do:

```sh
TMP_FILE="$(mktemp)"
if flameshot gui --path "${TMP_FILE}" 2&> /dev/null; then
    
    ## >>> your upload script for your preferred storage <<<
    
    rm "${TMP_FILE}"
fi
```

For further info about the CLI, you can always read the manpage, the `--help` for each subcommand, or visit https://flameshot.org/docs/advanced/commandline-options/

> Is a plug-in system already available or is this planned for he future? e.g. something like sharex plugin files.

there is already #3661 for the plugin system, but it it is not merged and I believe it needs some work as it was done as part of a commission work for a company by one of our contributors. I don't think it will get merged in version 13 of Flameshot, but hopefully it will be merged (@borgmanJeremy what do you think?)

**Reply by borgmanJeremy** [reply]
@mmahmoudian  I commented directly on the plugin PR thanks for the reminder. I don't think its ready for merging.

From the image uploader standpoint I __really__ want to yank it out of Flameshot. It feels like an entirely separate tool that is currently bolted on, and as many users have stated it's a privacy nightmare. I know this is an unpopular opinion which is why I haven't just removed it.

As you stated above uploading is already easily handled via the CLI since we support `--raw` so you can chain the screenshot to any other tool you want. This IMO is the right way to do it. Flameshot is the tool for screenshots / annotations which can be chained into another tool for uploading.

**Reply by tessus** [reply]
@mmahmoudian 

> I'm also a fan of your work in Joplin :) Thanks for your contributions in that FLOSS project.

Wow, thanks. Although I haven't contributed to Joplin lately, except merging translation PRs. All the code that I wrote was pretty much in the first 2 or 3 years of the project.

> Flameshot has [a very good CLI](https://flameshot.org/docs/advanced/commandline-options/) for macOS and Linux

Ahh, I see. Hmm, but this means I have to start it from the cli. The issue with that is that when I am in a terminal, I don't use screenshots, but rather use the terminal's logging options or asciinema. Also the terminal would be in the way. I basically require the app and thus also the script to be started from a global shortcut. (Maybe I can look into how to start a script with a global shortcut on macOS. I believe to recall that this is very clunky. I am great with bash scripts, but suck with Apple scripts.)

This is why calling a script from the app (e.g. `Cmd+U` for upload script) would be awesome.
I saw in the README that it is possible to open the capture in another app.
Unfortunatelty this shortcut does not work and there is no button for that either. Otherwise I could just use this workaround to send the capture to a script.

> there is already https://github.com/flameshot-org/flameshot/pull/3661 for the plugin system, but it it is not merged and I believe it needs some work as it was done as part of a commission work for a company by one of our contributors. I

Nice, fingers crossed.

@borgmanJeremy 

> uploading is already easily handled via the CLI since we support --raw so you can chain the screenshot to any other tool you want. 

However, this requires to start flameshot from the cli, which is not the usual workflow. The workflow is the other way around. 

So how exactly are you doing this? You work in the GUI and want to quickly take a screenshot, so you open a terminal, make it really, really small, so that it is not in the way of what you want to capture and then you start the script that opens the app to take the screenshot?
This is even slower than to just save the image into a dir and then upload it manually.
I am sorry, I just don't understand, but maybe I am missing something or don't understand the script workflow.

**Reply by borgmanJeremy** [reply]
@tessus I think the main difference is my desktop is 99% terminal with tmux haha. So  I often call flameshot from the terminal.

I think the best resolution for all of this is to figure out the plugin system and then all uploaders can be implemented as plugins outside the core program. This satisfies your use case and alleviates the privacy concerns I have.

**Reply by tessus** [reply]
I am also using the terminal extensively, but I dislike screenshots of text with the heat of a thousand suns. I especially adore it, if people send me a screenshot of a cut off error message, wrapped in a Word document. This just makes my day... every single time. ;-)

**Reply by mmahmoudian** [reply]
@tessus 
> So how exactly are you doing this? You work in the GUI and want to quickly take a screenshot, so you open a terminal, make it really, really small, so that it is not in the way of what you want to capture and then you start the script that opens the app to take the screenshot?

I haven't used macOS for about 10 years, and back then my workflow was very minimal. But in Linux I add my scripts to the launcher or bound them to global keybindings. For example I have this dmenu script for screenshot and the  post processing:

https://codeberg.org/mehrad/dmenu_shot

For macOS it seems to be possible too:

https://apple.stackexchange.com/questions/224394/how-to-make-a-mac-os-x-app-with-a-shell-script

But if I want to run something from terminal and hide the terminal, I would simply do `flameshot gui --delay=3000` to have 3 seconds time to minimize the terminal.

But I also agree that having the plugin system ready can relieve some of the pressure from Flameshot devs and at the same time give more freedom to users.

**Comment by nosovk** [comment]
Actually it would be nice to have an option to host images on your own domain. Like: https://transloadit.com/devtips/creating-a-free-image-cdn-with-cloudflare-r2/
We can achieve that by mimicking of an imgur api, just need to add an option to put API host in settings. Then users will be able to put they own domain and key, and get ad-free images on their own domain
